components/renku_data_services/authz/authz.py

```diff
@@ -322,26 +374,54 @@ async def members(
         return members
 
     @staticmethod
-    def project_change(op: ProjectAuthzOperation):
-        """A decorator that updates the authorization database for different types of project operations."""
+    def authz_change(op: AuthzOperation, resource: ResourceType):
+        """A decorator that updates the authorization database for different types of operations."""
 
         class WithAuthz(Protocol):
             @property
             def authz(self) -> Authz: ...
 
-        def decorator(f: Callable[..., Awaitable[Project | ProjectUpdate]]):
+        def _extract_user_from_args(*args, **kwargs) -> base_models.APIUser:
+            potential_user: base_models.APIUser | None = None
+            if len(args) == 0:
+                potential_user = kwargs.get("user")
+                if not potential_user:
+                    potential_user = kwargs.get("requested_by")
```

Comment:
Yes the logic here was a bit weird. It should never find both of those and have both of those be of `APIUser` type.

I added some code that will fail if this happens. The `APIUser` type is used only to represent a user that initiated the request. When a user is being edited, created or updated the payload is of a different type. Basically this code now will look for `APIUser` in both the `user` and `requested_by` keywords. And it will fail if it finds both and both are of type `APIUser`. If it only finds one of type `APIUser` then it will just use that value.